### ISSUE TITLE
Creating a typescript npm package using tscc

### ISSUE THREAD

**Issue by speakwithalisp** [theseanl/tscc#201]
Absolute noob here regarding closure compiler. I want to create a library, but I just can't get tscc to respect any of the names I want to export. The code is up at [https://github.com/speakwithalisp/csp-with-ts](https://github.com/speakwithalisp/csp-with-ts).

The (currently broken) package is 'csp-with-ts'.

Directory structure:

src
- index.d.ts
- index.ts
- impl
    - buffers.ts
     - channels.ts
     - constants.ts
     - go.ts
     - instructions.ts
     - interfaces.ts
     - process.ts
     - processEvents.ts
     - processQueue.ts
     - scheduler.ts
     - service.ts
     - utils.ts

tsconfig.json
```
{
    "include": ["src"],
    "files": ["src/index.ts",
              "src/impl/interfaces.ts",
              "src/impl/buffers.ts",
              "src/impl/channels.ts",
              "src/impl/constants.ts",
              "src/impl/go.ts",
              "src/impl/instructions.ts",
              "src/impl/process.ts",
              "src/impl/processEvents.ts",
              "src/impl/processQueue.ts",
              "src/impl/scheduler.ts",
              "src/impl/service.ts",
              "src/impl/utils.ts",
              "src/index.d.ts"
             ],
    "compilerOptions": {
        "strict": true,
        "strictNullChecks": true,
        "noImplicitThis": true,
        "downlevelIteration": true,
        "resolveJsonModule": true,
        "esModuleInterop": true,
        "moduleResolution": "node",
        "importHelpers": true,
        "module": "commonjs",
        "sourceMap": true,
        "forceConsistentCasingInFileNames": true,
        "allowSyntheticDefaultImports": true,
        "isolatedModules": true,
        "lib": [
            "dom",
            "dom.iterable",
            "esnext"
        ],
        "target": "es6"
    }
}
```

tscc.spec.json:
```
{
    "prefix": "lib/",
    "modules": {
        "index":   "src/index.ts"
    }
}

```

current output file:
```
function n(a){var d=0;return function(){return d<a.length?{done:!1,value:a[d++]}:{done:!0}}}var p="function"==typeof Object.defineProperties?Object.defineProperty:function(a,d,g){a!=Array.prototype&&a!=Object.prototype&&(a[d]=g.value)},q="undefined"!=typeof window&&window===this?this:"undefined"!=typeof global&&null!=global?global:this;function r(){r=function(){};q.Symbol||(q.Symbol=u)}function v(a,d){this.a=a;p(this,"description",{configurable:!0,writable:!0,value:d})}v.prototype.toString=function(){return this.a};
var u=function(){function a(g){if(this instanceof a)throw new TypeError("Symbol is not a constructor");return new v("jscomp_symbol_"+(g||"")+"_"+d++,g)}var d=0;return a}();function w(){r();var a=q.Symbol.iterator;a||(a=q.Symbol.iterator=q.Symbol("Symbol.iterator"));"function"!=typeof Array.prototype[a]&&p(Array.prototype,a,{configurable:!0,writable:!0,value:function(){return x(n(this))}});w=function(){}}function x(a){w();a={next:a};a[q.Symbol.iterator]=function(){return this};return a}
function y(a){var d="undefined"!=typeof Symbol&&Symbol.iterator&&a[Symbol.iterator];return d?d.call(a):{next:n(a)}}var z;if("function"==typeof Object.setPrototypeOf)z=Object.setPrototypeOf;else{var A;a:{var B={m:!0},C={};try{C.__proto__=B;A=C.m;break a}catch(a){}A=!1}z=A?function(a,d){a.__proto__=d;if(a.__proto__!==d)throw new TypeError(a+" is not extensible");return a}:null}var D=z;function E(){this.h=!1;this.b=null;this.j=void 0;this.a=1;this.l=0;this.g=null}
function F(a){if(a.h)throw new TypeError("Generator is already running");a.h=!0}E.prototype.i=function(a){this.j=a};function G(a,d){a.g={o:d,s:!0};a.a=a.l}E.prototype.return=function(a){this.g={return:a};this.a=this.l};function H(a){this.a=new E;this.b=a}function I(a,d){F(a.a);var g=a.a.b;if(g)return J(a,"return"in g?g["return"]:function(l){return{value:l,done:!0}},d,a.a.return);a.a.return(d);return K(a)}
function J(a,d,g,l){try{var h=d.call(a.a.b,g);if(!(h instanceof Object))throw new TypeError("Iterator result "+h+" is not an object");if(!h.done)return a.a.h=!1,h;var m=h.value}catch(f){return a.a.b=null,G(a.a,f),K(a)}a.a.b=null;l.call(a.a,m);return K(a)}function K(a){for(;a.a.a;)try{var d=a.b(a.a);if(d)return a.a.h=!1,{value:d.value,done:!1}}catch(g){a.a.j=void 0,G(a.a,g)}a.a.h=!1;if(a.a.g){d=a.a.g;a.a.g=null;if(d.s)throw d.o;return{value:d.return,done:!0}}return{value:void 0,done:!0}}
function L(a){this.next=function(d){F(a.a);a.a.b?d=J(a,a.a.b.next,d,a.a.i):(a.a.i(d),d=K(a));return d};this.throw=function(d){F(a.a);a.a.b?d=J(a,a.a.b["throw"],d,a.a.i):(G(a.a,d),d=K(a));return d};this.return=function(d){return I(a,d)};w();this[Symbol.iterator]=function(){return this}}function M(a,d){d=new L(new H(d));D&&D(d,a.prototype);return d}function N(a,d){return Object.prototype.hasOwnProperty.call(a,d)}
function O(a,d){if(d){var g=q;a=a.split(".");for(var l=0;l<a.length-1;l++){var h=a[l];h in g||(g[h]={});g=g[h]}a=a[a.length-1];l=g[a];d=d(l);d!=l&&null!=d&&p(g,a,{configurable:!0,writable:!0,value:d})}}
O("WeakMap",function(a){function d(b){this.a=(c+=Math.random()+1).toString();if(b){b=y(b);for(var e;!(e=b.next()).done;)e=e.value,this.set(e[0],e[1])}}function g(){}function l(b){var e=typeof b;return"object"===e&&null!==b||"function"===e}function h(b){if(!N(b,f)){var e=new g;p(b,f,{value:e})}}function m(b){var e=Object[b];e&&(Object[b]=function(k){if(k instanceof g)return k;h(k);return e(k)})}if(function(){if(!a||!Object.seal)return!1;try{var b=Object.seal({}),e=Object.seal({}),k=new a([[b,2],[e,
3]]);if(2!=k.get(b)||3!=k.get(e))return!1;k.delete(b);k.set(e,4);return!k.has(b)&&4==k.get(e)}catch(t){return!1}}())return a;var f="$jscomp_hidden_"+Math.random();m("freeze");m("preventExtensions");m("seal");var c=0;d.prototype.set=function(b,e){if(!l(b))throw Error("Invalid WeakMap key");h(b);if(!N(b,f))throw Error("WeakMap key fail: "+b);b[f][this.a]=e;return this};d.prototype.get=function(b){return l(b)&&N(b,f)?b[f][this.a]:void 0};d.prototype.has=function(b){return l(b)&&N(b,f)&&N(b[f],this.a)};
d.prototype.delete=function(b){return l(b)&&N(b,f)&&N(b[f],this.a)?delete b[f][this.a]:!1};return d});
O("Map",function(a){function d(){var c={};return c.f=c.next=c.head=c}function g(c,b){var e=c.a;return x(function(){if(e){for(;e.head!=c.a;)e=e.f;for(;e.next!=e.head;)return e=e.next,{done:!1,value:b(e)};e=null}return{done:!0,value:void 0}})}function l(c,b){var e=b&&typeof b;"object"==e||"function"==e?m.has(b)?e=m.get(b):(e=""+ ++f,m.set(b,e)):e="p_"+b;var k=c.b[e];if(k&&N(c.b,e))for(c=0;c<k.length;c++){var t=k[c];if(b!==b&&t.key!==t.key||b===t.key)return{id:e,list:k,index:c,c:t}}return{id:e,list:k,
index:-1,c:void 0}}function h(c){this.b={};this.a=d();this.size=0;if(c){c=y(c);for(var b;!(b=c.next()).done;)b=b.value,this.set(b[0],b[1])}}if(function(){if(!a||"function"!=typeof a||!a.prototype.entries||"function"!=typeof Object.seal)return!1;try{var c=Object.seal({x:4}),b=new a(y([[c,"s"]]));if("s"!=b.get(c)||1!=b.size||b.get({x:4})||b.set({x:4},"t")!=b||2!=b.size)return!1;var e=b.entries(),k=e.next();if(k.done||k.value[0]!=c||"s"!=k.value[1])return!1;k=e.next();return k.done||4!=k.value[0].x||
"t"!=k.value[1]||!e.next().done?!1:!0}catch(t){return!1}}())return a;w();var m=new WeakMap;h.prototype.set=function(c,b){c=0===c?0:c;var e=l(this,c);e.list||(e.list=this.b[e.id]=[]);e.c?e.c.value=b:(e.c={next:this.a,f:this.a.f,head:this.a,key:c,value:b},e.list.push(e.c),this.a.f.next=e.c,this.a.f=e.c,this.size++);return this};h.prototype.delete=function(c){c=l(this,c);return c.c&&c.list?(c.list.splice(c.index,1),c.list.length||delete this.b[c.id],c.c.f.next=c.c.next,c.c.next.f=c.c.f,c.c.head=null,
this.size--,!0):!1};h.prototype.clear=function(){this.b={};this.a=this.a.f=d();this.size=0};h.prototype.has=function(c){return!!l(this,c).c};h.prototype.get=function(c){return(c=l(this,c).c)&&c.value};h.prototype.entries=function(){return g(this,function(c){return[c.key,c.value]})};h.prototype.keys=function(){return g(this,function(c){return c.key})};h.prototype.values=function(){return g(this,function(c){return c.value})};h.prototype.forEach=function(c,b){for(var e=this.entries(),k;!(k=e.next()).done;)k=
k.value,c.call(b,k[1],k[0],this)};h.prototype[Symbol.iterator]=h.prototype.entries;var f=0;return h});r();w();r();w();r();w();r();w();function P(a,d){function g(){var m,f;return M(g,function(c){switch(c.a){case 1:m=h();f=m.next();c.a=2;break;case 3:return m.return(f.value),l=!0,c.return();case 2:if(f.done)return m.return(f.value),l=!0,c.return();case 4:if(f.done){c.a=5;break}var b=f.value;c.a=6;return{value:b};case 6:f=m.next();if(f.done){c.a=5;break}c.a=4;break;case 5:l=!0,c.a=0}})}var l=!1,h=d;Object.defineProperties(g.prototype,{isDone:{get:function(){return l}},channel:{get:function(){return a}},event:{get:function(){return 0}},
altFlag:{get:function(){return!1}}});return g()}
function Q(a,d,g,l){function h(){var f,c,b;return M(h,function(e){switch(e.a){case 1:f=d?d():void 0,c=!1,f&&f.next();case 2:if(c){e.a=3;break}e.a=4;return{value:void 0};case 4:b=e.j;if(void 0===b||null===b){f&&f.return();m=!0;e.a=3;break}if(l&&!l(b)){f&&f.return();m=!0;e.a=3;break}f&&(c=!!f.next(b).done);e.a=2;break;case 3:!c&&f&&f.return(),m=!0,e.a=0}})}var m=!1;Object.defineProperties(h.prototype,{isDone:{get:function(){return m}},channel:{get:function(){return a}},event:{get:function(){return 1}},
altFlag:{get:function(){return!!g}}});return h()};(function(){var a=new Map;r();a.set(Symbol.for(">!"),function(d){return P(this,d)});r();a.set(Symbol.for("<!"),function(d,g){return void 0!==g?Q(this,d,!1,g):Q(this,d)});return a})();
//# sourceMappingURL=index.js.map

```

**Comment by theseanl** [comment]
Hi, this requires some creativity on your end. I think you can consult how it is done in [fontoxpath](https://github.com/FontoXML/fontoxpath). Your entry file currently consists of lots of `export` statements. You may change it or create another entry point which insteads assigns every exported object as a property of the global `window` object, like in https://github.com/FontoXML/fontoxpath/blob/master/src/index.ts#L105. Then wrap it with an iife so that the `module.exports` object is fed to `window`, as in https://github.com/FontoXML/fontoxpath/blob/master/build.js#L82.

Another way to achieve it would be using `goog.exportSymbol` which is a closure compiler primitive. Check out its call signature in https://github.com/google/closure-library/blob/master/closure/goog/base.js#L1892. To use it, reference the declaration file [`goog.d.ts` ](https://github.com/theseanl/tscc/blob/master/packages/tscc/third_party/closure_library/base.d.ts) and write like
```ts
import goog from 'goog:goog';
goog.exportSymbol('loop', loop, module.exports)
```

**Comment by speakwithalisp** [comment]
Just wanted to update that goog.exportSymbol worked perfectly. Unfortunately, I have a lot of named properties for the internals, so sticking to the rollup tscc plugin is probably the path of least resistance for me.

The fontoxpath guys have a very interesting and comprehensive approach, and I think that seems to be the way to go if one wants to compile to a library (which is my exact use case). But the rollup plugin outputs decent enough code, and I'm already able to use my library in production so I'll probably not seriously pursue this approach.

Thanks a lot for the pointers, they really helped me understand the pretty arcane ways of the goog, and if I ever plan on making another library, I'll design it to be more closure compiler friendly (the rollup plugin already does the job perfectly for me though so I am using the closure compiler at some level I guess).

Sorry for the late update, I only got around to trying this out today.

P.S: 
Just a note for anyone who might be as inattentive as I am when reading comments, I struggled a bit with goog until I read 

> reference the declaration file goog.d.ts 

This means use the triple slash directives. The maintainer has a demo repo which documents this very well, just search for the backbone example.
